keps/sig-api-machinery/4153-declarative-validation/README.md

```diff
@@ -0,0 +1,1793 @@
+to get hands on experience with this proposed enhancement.
+
+### Goals
+
+- Vast majority (95%+) of hand written validation are replaced with declarative
+validation. The remaining hand written rules will primarily be rules that _should
+not_ be published as part of the API, usually because server side state is
+involved in the validation decision.
+
+- `types.go` files become the de-facto IDL of Kubernetes for native types.
+It is worthing noting that `+enum` support, `+default` support and similar
+enhancements all moved our API development forward in this direction. This
+enhancement is an attempt to finish that story arc.
+
+- CRDs and native types are validated and published in OpenAPI in a consistent
+and uniform way. Improvement we make to CEL to enable declarative validation of
+native types will be made available to CRD authors and vis-versa.
+
+- No change to error message structure for clients. Difference to clients is 
+limited to minimal changes to error detail strings.
+
+### Non-Goals
+
+- It is not a goal for _all_ validation rules to be published to OpenAPI. There
+  exist complex, stateful rules that should not be published in OpenAPI. These
+  will remain "server side only" validation rules.
+- It is not a goal to require that _all_ validation rules be written in CEL. We
+  will continue to support hand written validation rules indefinitely for use
+  cases that are a poor fit using the existing validation mechanism (or
+  something very similar).
+
+
+## Proposal
+
+Declarative validation will be performed against versioned APIs. This differs from the hand
+written validation, which is evaluated against internal types.
+
+Go IDL tags will be added to support the following declarative validation rules:
+
+| Type of valiation      | Go IDL tag                                                       | OpenAPI validation field                                                           | Reason Type      |
+| ---------------------- | ---------------------------------------------------------------- | ---------------------------------------------------------------------------------- | ---------------- |
+| string format          | `+format={format name}`                                          | `format`                                                                           | Invalid          |
+| size limits            | `+min{Length,Properties,Items}`, `+max{Length,Properties,Items}` | `min{Length,Properties,Items}`, `max{Length,Properties,Items}`                     | TooMany, TooLong |
+| numeric limits         | `+minimum`, `+maximum`, `+exclusiveMinimum`, `+exclusiveMaximum` | `minimum`, `maximum`, `exclusiveMinimum`, `exclusiveMaximum`                       | Invalid          |
+| required fields        | `+optional` (exists today)                                       | `required`                                                                         | Required         |
```

Comment:
Can we use this to formalize the relationship between optional and defaults?  IOW if a field is truly optional, it sets `+optional`.  If a field is required but has a default, it sets `+default` (or we can spec it as `+optional` AND `+default`).  If a field is required from the user it sets neither.

Reply:
Tim and I can't agree on what "optional"/"required" means because I look at it from a  kubectl applier's perspective (a defaulted field MUST be optional, how can it be defaulted if it was required in the first place?), and I think he looks at it more from a controller perspective (a defaulted field is required because it will always be present when you do a GET)

Reply:
Exactly.  When I say "formalize" I really mean "write down how people express the difference".  Today we only have +optional and +default is pretty new and not widely used.  Before it becomes the norm, let's just write down (API convention doc or something) which one (or both) should be used.

Reply:
Whatever we decide, seems like a good thing to add as an [APILinter rule in `kube-openapi`](https://github.com/kubernetes/kube-openapi/blob/master/pkg/generators/rules/names_match.go).

Personally, it feels natrual for me to think of the versioned schema should from user perspective. So any field with a `+default` must be `+optional`.

Reply:
The user perspective varies depending on whether you're the writer or the reader:

1. writer: defaulted fields are optional, I don't have to specify
2. reader: default fields are required, they will always have values.

fwiw, I personally find that defaulted fields as required matches my expectations, but allowing exactly one of required, optional, default maybe the clearest.